src/Compilers/CSharp/Portable/Binder/SwitchBinder.cs

```diff
@@ -205,7 +205,29 @@ private BoundExpression ConvertCaseExpression(TypeSymbol switchGoverningType, CS
 
             if (switchGoverningType.IsNullableType() && convertedCaseExpression.Kind == BoundKind.Conversion)
             {
-                constantValueOpt = ((BoundConversion)convertedCaseExpression).Operand.ConstantValue;
+                var underlyingType = switchGoverningType.GetNullableUnderlyingType();
+                var operand = ((BoundConversion)convertedCaseExpression).Operand;
+
+                if (operand.ConstantValue.IsIntegral)
+                {
+                    // We have to make sure we convert numerics correctly
+                    //public static bool F(long? x)
+                    //{
+                    //    switch (x)
+                    //    {
+                    //        // '1' has to be converted from Int32 to Int64 here
+                    //        case 1:
+                    //            return true;
+                    //        default:
+                    //            return false;
+                    //    }
+                    //}
+                    constantValueOpt = FoldConstantConversion(convertedCaseExpression.Syntax, operand, Conversion.ImplicitNumeric, underlyingType, diagnostics);
```

Comment:
It is inaccessible in this constant (it is private) and I did not want to change that